daemon/oci_linux.go

```diff
@@ -98,6 +100,31 @@ func setDevices(s *specs.Spec, c *container.Container) error {
 		for _, d := range hostDevices {
 			devs = append(devs, oci.Device(d))
 		}
+		for _, deviceMapping := range c.HostConfig.Devices {
+			if deviceMapping.PathOnHost == deviceMapping.PathInContainer {
+				logrus.WithField("container", c.ID).Warnf("Specifying a device %s is redundant for privileged mode", deviceMapping.PathOnHost)
+				if deviceMapping.CgroupPermissions != "rwm" {
+					logrus.WithField("container", c.ID).Warnf("Specifying permission %s to device %s is redundant for privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
+				}
+				continue
+			}
+
+			if _, err := os.Stat(deviceMapping.PathInContainer); err != nil {
+				if os.IsNotExist(err) {
+					if deviceMapping.CgroupPermissions != "rwm" {
+						logrus.WithField("container", c.ID).Warnf("Specifying permission %s to device %s is redundant for privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
+					}
+					d, _, error := oci.DevicesFromPath(deviceMapping.PathOnHost, deviceMapping.PathInContainer, "rwm")
+					if error != nil {
+						return error
+					}
+					devs = append(devs, d...)
+					continue
+				}
+				return errors.Wrap(err, "error stating device path in contianer")
+			}
+			return errdefs.InvalidParameter(errors.Errorf("container device path: %s must be different from any host device path for privileged mode containers", deviceMapping.PathInContainer))
```

Comment:
Same question as above: in what situation do we end here? 

The error message seems to be if `deviceMapping.PathOnHost == deviceMapping.PathInContainer` (but that's already handled above, and prints a _warning_ ? (perhaps I'm looking wrong, 😅)

Reply:
Change the logic to make it more clear